flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala

```diff
@@ -581,6 +581,62 @@ class JobManager(
           )
       }
 
+    case CancelJobWithSavepoint(jobId, savepointDirectory) =>
+      try {
+        val targetDirectory = if (savepointDirectory != null) {
+          savepointDirectory
+        } else {
+          defaultSavepointDir
+        }
+
+        log.info(s"Trying to cancel job $jobId with savepoint to $targetDirectory")
+
+        currentJobs.get(jobId) match {
+          case Some((executionGraph, _)) =>
+            // We don't want any checkpoint between the savepoint and cancellation
+            val coord = executionGraph.getCheckpointCoordinator
+            coord.stopCheckpointScheduler()
```

Comment:
I think it's not enough to simply call `stopCheckpointScheduler`. If I'm not mistaken, then the following could happen: You call `stopCheckpointScheduler` which will try to `cancel` the last `currentPeriodicTrigger`. Now assume that the last `TimerTask` to trigger the next checkpoint has just been triggered but not executed (just before cancelling it). Now the `stopCheckpointScheduler` finishes without the `TimerTask` having completed. Now the `TimerTask` can still trigger a checkpoint even though we've stopped the checkpoint scheduler.

The way to fix this (admittedly academic corner case), is to filter out outdated `TimerTask` calls in the `CheckpointCoordinator` by having a kind of fencing tokens for the trigger checkpoint calls.